crates/cubecl-runtime/src/client.rs

```diff
@@ -349,17 +383,71 @@ where
     }
 
     /// Measure the execution time of some inner operations.
-    ///
-    /// Nb: this function will only allow one function at a time to be submitted when multi threading.
-    /// Recursive measurements are not allowed and will deadlock.
-    pub fn profile(&self, func: impl FnOnce()) -> ProfileDuration {
+    #[track_caller]
+    pub fn profile(
+        &self,
+        func: impl FnOnce(),
+        #[allow(unused)] func_name: &str,
+    ) -> ProfileDuration {
+        // Get the outer caller. Maybe this should be lazy, idk.
+        // For execute() this could also maybe point straight to the
+        // cube kernel instead. For general profiling that's not possible.
+        #[cfg(feature = "profile-tracy")]
+        let location = std::panic::Location::caller();
+
+        // Make a CPU span. If the server has system profiling this is all you need.
+        #[cfg(feature = "profile-tracy")]
+        let _span = tracy_client::Client::running().unwrap().span_alloc(
+            None,
+            func_name,
+            location.file(),
+            location.line(),
+            0,
+        );
+
         #[cfg(multi_threading)]
-        let stream_id = self.profile_aquire();
-        let token = self.channel.start_profile();
+        let stream_id = self.profile_acquire();
 
-        func();
+        let token = self.channel.start_profile();
+        let method = self.state.properties.timing_method;
+
+        let result = if method == TimingMethod::Device {
+            #[cfg(feature = "profile-tracy")]
+            let mut gpu_span = self
+                .state
+                .gpu_client
+                .span_alloc(func_name, "profile", location.file(), location.line())
+                .unwrap();
+            func();
+            #[cfg(feature = "profile-tracy")]
+            gpu_span.end_zone();
+
+            #[allow(unused_mut)]
+            let mut result = self.channel.end_profile(token);
+
+            #[cfg(feature = "profile-tracy")]
+            {
+                let epoch = self.state.epoch_time;
+
+                // Add in the work to upload the timestamp data.
+                result = ProfileDuration::new(
+                    Box::pin(async move {
+                        let ticks = result.resolve().await;
+                        let start_duration = ticks.start_duration_since(epoch).as_nanos() as i64;
+                        let end_duration = ticks.end_duration_since(epoch).as_nanos() as i64;
+                        gpu_span.upload_timestamp_start(start_duration);
+                        gpu_span.upload_timestamp_end(end_duration);
+                        ticks
+                    }),
+                    method,
+                );
+            }
 
-        let result = self.channel.end_profile(token);
+            result
+        } else {
+            func();
+            self.channel.end_profile(token)
+        };
```

Comment:
Not sure what's the difference between the two blocks. When tracy is activated, it has to create a span and collect the result, but the `func` and `end_profile` are the same. Maybe put a single `cfg(profile-tracy)` above the block that add the span, and log a warning is device timing method isn't in used? Also is it necessary to have device profiling for tracy to work?